File: Service/Bracket/DoubleElimBracketBuilder.cs
namespace TecmoTourney.Bracket;

/// <summary>Server-side mirror of <c>double-elim-bracket.builder.ts</c>.</summary>
public static class DoubleElimBracketBuilder
{
    public static int BracketSizeForEntrantCount(int n)
    {
        if (n < 4)
            return 4;
        var b = 4;
        while (b < n && b < 32)
            b *= 2;
        if (b < n)
            b = 32;
        return b;
    }

    public static int[] BracketLeafRanks(int size)
    {
        if (size <= 1)
            return new[] { 0 };
        var half = size / 2;
        var prev = BracketLeafRanks(half);
        var result = new int[size];
        var idx = 0;
        for (var i = 0; i < half; i++)
        {
            result[idx++] = prev[i];
            result[idx++] = size - 1 - prev[i];
        }
        return result;
    }

    private static int LbRoundMatchCount(int b, int j)
    {
        if (j == 0)
            return b / 4;
        if (j % 2 == 1)
            return (int)(b / Math.Pow(2, (j + 1) / 2 + 1));
        return (int)(b / Math.Pow(2, j / 2 + 2));
    }

    private static FeedRef WbLeafFeed(int n, int[] leafRanks, int leafIndex)
    {
        var rank = leafRanks[leafIndex];
        if (rank >= n)
            return new FeedRef { Kind = FeedKind.Bye };
        return new FeedRef { Kind = FeedKind.Seed, SeedSlot = rank };
    }

    public static (int B, List<BracketMatch> Matches) BuildDoubleEliminationMatches(int entrantCount)
    {
        var n = entrantCount;
        var b = BracketSizeForEntrantCount(n);
        var k = (int)Math.Log2(b);
        var leafRanks = BracketLeafRanks(b);
        var matches = new List<BracketMatch>();

        for (var r = 0; r < k; r++)
        {
            var count = (int)(b / Math.Pow(2, r + 1));
            for (var i = 0; i < count; i++)
            {
                var top = r == 0
                    ? WbLeafFeed(n, leafRanks, 2 * i)
                    : new FeedRef { Kind = FeedKind.Winner, MatchId = $"wb-{r - 1}-{2 * i}" };
                var bottom = r == 0
                    ? WbLeafFeed(n, leafRanks, 2 * i + 1)
                    : new FeedRef { Kind = FeedKind.Winner, MatchId = $"wb-{r - 1}-{2 * i + 1}" };
                matches.Add(new BracketMatch
                {
                    Id = $"wb-{r}-{i}",
                    Segment = "WB",
                    Round = r,
                    IndexInRound = i,
                    Top = top,
                    Bottom = bottom
                });
            }
        }

        var lbRounds = 2 * k - 2;
        for (var j = 0; j < lbRounds; j++)
        {
            var count = LbRoundMatchCount(b, j);
            for (var i = 0; i < count; i++)
            {
                FeedRef top;
                FeedRef bottom;
                if (j == 0)
                {
                    top = new FeedRef { Kind = FeedKind.Loser, MatchId = $"wb-0-{2 * i}" };
                    bottom = new FeedRef { Kind = FeedKind.Loser, MatchId = $"wb-0-{2 * i + 1}" };
                }
                else if (j % 2 == 1)
                {
                    var wbRound = (j + 1) / 2;
                    // Winner of previous LB round match i (not 2*i — e.g. j=1,i=1 must be lb-0-1, not lb-0-2).
                    top = new FeedRef { Kind = FeedKind.Winner, MatchId = $"lb-{j - 1}-{i}" };
                    bottom = new FeedRef { Kind = FeedKind.Loser, MatchId = $"wb-{wbRound}-{i}" };
                }
                else
                {
                    top = new FeedRef { Kind = FeedKind.Winner, MatchId = $"lb-{j - 1}-{2 * i}" };
                    bottom = new FeedRef { Kind = FeedKind.Winner, MatchId = $"lb-{j - 1}-{2 * i + 1}" };
                }

                matches.Add(new BracketMatch
                {
                    Id = $"lb-{j}-{i}",
                    Segment = "LB",
                    Round = j,
                    IndexInRound = i,
                    Top = top,
                    Bottom = bottom
                });
            }
        }

        matches.Add(new BracketMatch
        {
            Id = "gf-0-0",
            Segment = "GF",
            Round = 0,
            IndexInRound = 0,
            Top = new FeedRef { Kind = FeedKind.Winner, MatchId = $"wb-{k - 1}-0" },
            Bottom = new FeedRef { Kind = FeedKind.Winner, MatchId = $"lb-{lbRounds - 1}-0" }
        });

        matches.Add(new BracketMatch
        {
            Id = "gf-1-0",
            Segment = "GF",
            Round = 1,
            IndexInRound = 0,
            Top = new FeedRef { Kind = FeedKind.Empty },
            Bottom = new FeedRef { Kind = FeedKind.Empty }
        });

        return (b, matches);
    }
}

File: Client/tecmo-tourney/src/app/features/tournaments/bracket/double-elim-bracket.builder.ts
import { BracketMatch, FeedRef } from './double-elim-bracket.types';

/** Next power of two in [4, 32]. */
export function bracketSizeForEntrantCount(n: number): number {
  if (n < 4) return 4;
  let b = 4;
  while (b < n && b < 32) b *= 2;
  if (b < n) b = 32;
  return b;
}

/** Classic tournament leaf ordering: seed ranks 0..B-1 at WB round-1 leaves (0 = best). */
export function bracketLeafRanks(size: number): number[] {
  if (size <= 1) return [0];
  const half = size / 2;
  const prev = bracketLeafRanks(half);
  const result: number[] = [];
  for (let i = 0; i < half; i++) {
    result.push(prev[i]);
    result.push(size - 1 - prev[i]);
  }
  return result;
}

function lbRoundMatchCount(B: number, j: number): number {
  if (j === 0) return B / 4;
  if (j % 2 === 1) {
    return B / Math.pow(2, (j + 1) / 2 + 1);
  }
  return B / Math.pow(2, j / 2 + 2);
}

function wbLeafFeed(N: number, B: number, leafIndex: number, leafRanks: number[]): FeedRef {
  const rank = leafRanks[leafIndex];
  if (rank >= N) return { kind: 'bye' };
  return { kind: 'seed', seedSlot: rank };
}

export function buildDoubleEliminationMatches(N: number): { B: number; matches: BracketMatch[] } {
  const B = bracketSizeForEntrantCount(N);
  const k = Math.log2(B);
  const leafRanks = bracketLeafRanks(B);
  const matches: BracketMatch[] = [];

  for (let r = 0; r < k; r++) {
    const count = B / Math.pow(2, r + 1);
    for (let i = 0; i < count; i++) {
      const top: FeedRef =
        r === 0
          ? wbLeafFeed(N, B, 2 * i, leafRanks)
          : { kind: 'winner', matchId: `wb-${r - 1}-${2 * i}` };
      const bottom: FeedRef =
        r === 0
          ? wbLeafFeed(N, B, 2 * i + 1, leafRanks)
          : { kind: 'winner', matchId: `wb-${r - 1}-${2 * i + 1}` };
      matches.push({
        id: `wb-${r}-${i}`,
        segment: 'WB',
        round: r,
        indexInRound: i,
        top,
        bottom
      });
    }
  }

  const lbRounds = 2 * k - 2;
  for (let j = 0; j < lbRounds; j++) {
    const count = lbRoundMatchCount(B, j);
    for (let i = 0; i < count; i++) {
      let top: FeedRef;
      let bottom: FeedRef;
      if (j === 0) {
        top = { kind: 'loser', matchId: `wb-0-${2 * i}` };
        bottom = { kind: 'loser', matchId: `wb-0-${2 * i + 1}` };
      } else if (j % 2 === 1) {
        const wbRound = (j + 1) / 2;
        // Winner of previous LB round match i (not 2*i — e.g. j=1,i=1 must be lb-0-1, not lb-0-2).
        top = { kind: 'winner', matchId: `lb-${j - 1}-${i}` };
        bottom = { kind: 'loser', matchId: `wb-${wbRound}-${i}` };
      } else {
        top = { kind: 'winner', matchId: `lb-${j - 1}-${2 * i}` };
        bottom = { kind: 'winner', matchId: `lb-${j - 1}-${2 * i + 1}` };
      }
      matches.push({
        id: `lb-${j}-${i}`,
        segment: 'LB',
        round: j,
        indexInRound: i,
        top,
        bottom
      });
    }
  }

  matches.push({
    id: 'gf-0-0',
    segment: 'GF',
    round: 0,
    indexInRound: 0,
    top: { kind: 'winner', matchId: `wb-${k - 1}-0` },
    bottom: { kind: 'winner', matchId: `lb-${lbRounds - 1}-0` }
  });

  matches.push({
    id: 'gf-1-0',
    segment: 'GF',
    round: 1,
    indexInRound: 0,
    top: { kind: 'empty' },
    bottom: { kind: 'empty' }
  });

  return { B, matches };
}
